packages/web/src/hooks/useExternalWalletSwap.ts

```diff
@@ -1,44 +1,261 @@
+import {
+  QuoteResponse,
+  SwapInstructionsResponse,
+  SwapRequest
+} from '@jup-ag/api'
 import type { Provider as SolanaProvider } from '@reown/appkit-adapter-solana/react'
-import { VersionedTransaction } from '@solana/web3.js'
+import {
+  PublicKey,
+  TransactionInstruction,
+  TransactionMessage,
+  VersionedTransaction
+} from '@solana/web3.js'
 import { useMutation, useQueryClient } from '@tanstack/react-query'
 
 import { appkitModal } from 'app/ReownAppKitModal'
 import { reportToSentry } from 'store/errors/reportToSentry'
 
-type ExternalWalletSwapParams = {
-  inputAmountUi: number
-  inputToken: { decimals: number; address: string }
-  outputToken: { decimals: number; address: string }
+export type ExternalWalletSwapParams = {
+  inputDecimals: number
+  outputDecimals: number
+  walletAddress: string
+} & SwapTokensParams
+
+type IndirectSwapParams = {
+  inputMint: string
+  outputMint: string
+  audioMint: string
+  inputDecimals: number
+  outputDecimals: number
+  audioDecimals: number
+  amountUi: number
   walletAddress: string
-  isAMM: boolean
+  solanaConnection: any
+}
+
+const getIndirectSwapTx = async ({
+  inputMint,
+  outputMint,
+  audioMint,
+  inputDecimals,
+  outputDecimals,
+  audioDecimals,
+  amountUi,
+  walletAddress,
+  solanaConnection
+}: IndirectSwapParams): Promise<{
+  transaction: VersionedTransaction
+  inputAmount: { amount: number; uiAmount: number }
+  outputAmount: { amount: number; uiAmount: number }
+}> => {
+  // Get quote for first hop: input -> AUDIO
+  const { quoteResult: firstQuote } = await getJupiterQuoteByMintWithRetry({
+    inputMint,
+    outputMint: audioMint,
+    inputDecimals,
+    outputDecimals: audioDecimals,
+    amountUi,
+    swapMode: 'ExactIn',
+    onlyDirectRoutes: false
+  })
+
+  // Use the output of first swap as input for second swap
+  const audioAmount = firstQuote.outputAmount.uiAmount
+
+  // Get quote for second hop: AUDIO -> output
+  const { quoteResult: secondQuote } = await getJupiterQuoteByMintWithRetry({
+    inputMint: audioMint,
+    outputMint,
+    inputDecimals: audioDecimals,
+    outputDecimals,
+    amountUi: audioAmount,
+    swapMode: 'ExactIn',
+    onlyDirectRoutes: false
+  })
+
+  // Get instructions for both swaps
+  const swapRequest1: SwapRequest = {
+    quoteResponse: firstQuote.quote,
+    userPublicKey: walletAddress,
+    dynamicSlippage: true,
+    useSharedAccounts: false
+  }
+
+  const swapRequest2: SwapRequest = {
+    quoteResponse: secondQuote.quote,
+    userPublicKey: walletAddress,
+    dynamicSlippage: true,
+    useSharedAccounts: false
+  }
+
+  let firstSwapInstructions: SwapInstructionsResponse
+  let secondSwapInstructions: SwapInstructionsResponse
+
+  try {
+    firstSwapInstructions = await jupiterInstance.swapInstructionsPost({
+      swapRequest: swapRequest1
+    })
+  } catch (e) {
+    // Retry without shared accounts if it fails
+    swapRequest1.useSharedAccounts = false
```

Comment:
it's already false. also, we try 3 more times in internal flow, can perhaps try more times here?